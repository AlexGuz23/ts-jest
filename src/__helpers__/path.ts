import { resolve, relative } from 'path'

export const ROOT = resolve(__dirname, '..', '..')

export function relativeToRoot(path: string) {
  return relative(ROOT, path)
}
