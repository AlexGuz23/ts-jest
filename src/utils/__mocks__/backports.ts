import * as _backports from '../backports';

export const backportJestConfig: typeof _backports['backportJestConfig'] = val => ({
  ...(val as any),
});
