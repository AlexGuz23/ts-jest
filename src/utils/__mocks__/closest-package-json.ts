import _closestPackageJson from '../closest-package-json';
import { TPackageJson } from '../../types';

// tslint:disable-next-line:variable-name
export const __byPath: Record<string, TPackageJson> = {};
// tslint:disable-next-line:variable-name
export const __default: TPackageJson = { name: 'mock' };

const closestPackageJson: typeof _closestPackageJson = (
  fromPath: string,
  asString: boolean = false,
) => {
  const res = fromPath in __byPath ? __byPath[fromPath] : __default;
  return (asString ? JSON.stringify(res) : res) as any;
};

export default closestPackageJson;
