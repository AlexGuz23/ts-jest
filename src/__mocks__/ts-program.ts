import * as fakers from '../__helpers__/fakers'
import { TsJestConfig, TsJestProgram } from '../types'
import { ParsedCommandLine } from 'typescript'

// tslint:disable-next-line:variable-name
export let __tsConfig: any = {}

export default class TsProgramMock implements TsJestProgram {
  get parsedConfig(): ParsedCommandLine {
    return __tsConfig
  }

  constructor(
    public rootDir: string = fakers.filePath(''),
    public tsJestConfig: TsJestConfig = fakers.tsJestConfig(),
  ) {}

  transpileModule(_: string, source: string) {
    return source
  }
}
