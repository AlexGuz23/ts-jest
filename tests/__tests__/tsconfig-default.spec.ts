import { MockedPath } from '../__mocks__/path';
jest.mock('path');

import * as ts from 'typescript';
import getTSConfig from '../../dist/utils/get-ts-config';
import * as path from 'path';
import mockJestConfig from '../__helpers__/mock-jest-config';
import getTSJestConfig from '../../dist/utils/get-ts-jest-config';

const TEST_CASE = 'tsconfig-test';

describe('get default ts config', () => {
  beforeEach(() => {
    // Set up some mocked out file info before each test
    ((path as any) as MockedPath).__setBaseDir('./tests/tsconfig-test');
    // empties the lodash memoized cache for this function
    getTSConfig.cache.clear();
  });

  it('should correctly read tsconfig.json', () => {
    const result = getTSConfig(mockJestConfig(TEST_CASE));

    expect(result).toMatchSnapshot();
  });

  describe('new behavior (tsConfigFile & tsConfig)', () => {
    it('should be same results for null/undefined/etc.', () => {
      const resultWithoutTsJestSection = getTSConfig(
        mockJestConfig(TEST_CASE, null),
      );
      const resultEmptyParam = getTSConfig(mockJestConfig(TEST_CASE, {}));
      const resultUndefinedContentFile = getTSConfig(
        mockJestConfig(TEST_CASE, { tsConfigFile: undefined }),
      );
      const resultNullContentFile = getTSConfig(
        mockJestConfig(TEST_CASE, { tsConfigFile: null }),
      );

      expect(resultEmptyParam).toEqual(resultWithoutTsJestSection);
      expect(resultUndefinedContentFile).toEqual(resultWithoutTsJestSection);
      expect(resultNullContentFile).toEqual(resultWithoutTsJestSection);
    });

    it('should be different results for different rootDir with same jest config.', () => {
      const rootConfig = getTSConfig(mockJestConfig(TEST_CASE));
      const subConfig = getTSConfig(
        mockJestConfig(`${TEST_CASE}/tsconfig-module`),
      );
      expect(rootConfig).not.toEqual(subConfig);
    });

    it('should not change the module if it is loaded from a non-default config file', () => {
      const config = getTSConfig(
        mockJestConfig(TEST_CASE, {
          tsConfigFile: 'tsconfig-module/custom-config.json',
        }),
      );

      // snapshot would be enough here, but that adds a security in case we do not see it in a PR
      expect(config.module).toBe(ts.ModuleKind.ES2015);
      expect(config.module).toMatchSnapshot();
    });

    it('should set the module to CommonJS if it is not, when loading from the default tsconfig file', () => {
      // set the base directory such that we can use 'tsconfig.json' as the
      // config file name instead of 'dir/tsconfig.json'
      ((path as any) as MockedPath).__setBaseDir(
        './tests/tsconfig-test/tsconfig-module',
      );

      const config = getTSConfig(
        mockJestConfig(`${TEST_CASE}/tsconfig-module`),
      );

      expect(config.module).toBe(ts.ModuleKind.CommonJS);
    });
  });

  it('should correctly read the ts-jest object within jest settings', () => {
    ((path as any) as MockedPath).__setBaseDir(
      './tests/tsconfig-test/tsconfig-module',
    );

    const config1 = getTSJestConfig({
      globals: {
        'ts-jest': {
          skipBabel: true,
        },
      },
    });

    expect(config1.skipBabel).toBe(true);

    const config2 = getTSJestConfig({
      globals: {
        'ts-jest': {
          skipBabel: false,
        },
      },
    });

    expect(config2.skipBabel).toBe(false);

    expect(getTSJestConfig({ globals: { 'ts-jest': {} } })).toEqual({});
    expect(getTSJestConfig({})).toEqual({});
  });
});
