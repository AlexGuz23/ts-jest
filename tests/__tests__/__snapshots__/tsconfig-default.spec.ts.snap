// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get default ts config new behavior (tsConfigFile & tsConfig) should not change the module if it is loaded from a non-default config file 1`] = `5`;

exports[`get default ts config should correctly read tsconfig.json 1`] = `
Object {
  "configFilePath": undefined,
  "declaration": false,
  "declarationMap": false,
  "emitDeclarationOnly": false,
  "inlineSourceMap": true,
  "inlineSources": true,
  "jsx": 2,
  "module": 1,
  "moduleResolution": 2,
  "noEmitOnError": false,
  "target": 2,
}
`;
